feat(linter): no-unused-vars add setting for `reportVarsOnlyUsedAsTypes`

camc314 opened the pull request:
closes #11006

adds an option called `reportVarsOnlyUsedAsType`, defaulting to false

when enabled, it will report code such as

```ts
const x = 123;
export type X = typeof x;
```

---

Edit by Boshen:

This goes against typescript-eslint, where it reports `x` is unused by default.

* https://typescript-eslint.io/rules/no-unused-vars/#why-does-this-rule-report-variables-used-only-for-types
* https://github.com/typescript-eslint/typescript-eslint/issues/10266
* https://github.com/typescript-eslint/typescript-eslint/pull/10690

I think this will give better out-of-the box experience. It does not meet the criteria of `correctness` rules (rules are incorrect or useless).

In most situations, it is really hard for people to refactor their code when they see many `'mySchema' is defined but only used as a type.`:

```
import { z } from "zod";

const mySchema = z.object({ value: z.number() });

export type Shape = z.infer<typeof mySchema>;
```

camc314 commented:
* **#11009** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11009?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/11009?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=11009).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented on crates/oxc_linter/src/rules/eslint/no_unused_vars/usage.rs:
````diff
@@ -137,11 +137,25 @@ impl<'a> Symbol<'_, 'a> {
+                // const foo = 123;
+                // export type Foo = typeof foo
+                // ```
+                if options.report_vars_only_used_as_types
+                    && !self.flags().intersects(SymbolFlags::TypeImport.union(SymbolFlags::Import))
                     && self.reference_contains_type_query(reference)
                 {
                     continue;
````
incorrect i think? We are currently looping over all refernces of a declaration. If one of these has usages, we return true, else we continue to the next reference.

in this if block, we are checking if the reference is only used as a type, if it is we basically treat if as if it's unused

Boshen commented:
### Merge activity

* **May 13, 8:39 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **May 13, 8:39 PM EDT**: `Boshen` added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

camc314 committed:
feat(linter): no-unused-vars add setting for `reportVarsOnlyUsedAsTypes` (#11009)

closes #11006

adds an option called `reportVarsOnlyUsedAsType`, defaulting to false

when enabled, it will report code such as

```ts
const x = 123;
export type X = typeof x;
```
